src/odemis/acq/acqmng.py

```diff
@@ -83,6 +84,189 @@ def acquire(streams, settings_obs=None):
+        # cancel the sub-futures
+        self._actuator_f.cancel()
+        self._single_acqui_f.cancel()
+        return True
+
+    def _estimate_zstep_duration(self):
+        """
+        return (float > 0): estimated time (in s) that it takes to move the focus actuator
+          by one step.
+        """
+        # get the focuser of any of the FM streams
+        focuser = next(
+            (s.focuser for s in self._streams if isinstance(s, FluoStream)), None
+        )
+        speed = None
+        if focuser is not None:
+            if model.hasVA(focuser, "speed"):
+                speed = focuser.speed.value.get("z", None)
+        if speed is None:
+            speed = 10e-6  # m/s, pessimistic
+        zstep = 0
+        if self._zpos:
+            zstep = abs(self._zpos[0] - self._zpos[1])
+        return estimateMoveDuration(zstep, speed, 0.01)
+
+    def estimate_total_duration(self):
+        """
+        Estimates the total time for the zstack acquisition.
+        return (Float > 0): the estimated time
+        """
+        acqui_time = 0.0
+        zstep_time = 0
+        nr_FM_streams = 0
+        for s in self._streams:
+            acqui_time += s.estimateAcquisitionTime() * len(self._zlevels[s])
+            if isinstance(s, FluoStream):
+                nr_FM_streams += 1
+        if self._zpos:
+            zstep_time = self._zstep_duration * (len(self._zpos) - 1) * nr_FM_streams
+        return acqui_time + zstep_time
+
+    def _construct_zcube(self, data_arrays):
+        """
+        Create 3D data array from the 2D acquired data arrays 
+        data_arrays (list of DataArray): a list of acquired images at different z levels.
+        return (DataArray): a data array embedding all the data acquired at all z levels.
+        """
+        # pixel size along z axis
+        metadata = data_arrays[0].metadata
+        px, py = metadata[model.MD_PIXEL_SIZE]
+        pz = (self._zpos[-1] - self._zpos[0]) / (len(self._zpos) - 1)
+
+        if pz < 0:
+            data_arrays = numpy.flipud(data_arrays)
+            pz = -pz
+        # upgrade the metadata with the z pixel size
+        metadata[model.MD_PIXEL_SIZE] = (px, py, round(pz, ndigits=7))
+
+        # z coordinate of center
+        cx, cy = metadata[model.MD_POS]
+        cz = (self._zpos[0] + self._zpos[-1]) / 2
+        # upgrade the metadata with the z center position
+        metadata[model.MD_POS] = (cx, cy, round(cz, ndigits=5))
+
+        return model.DataArray(data_arrays, metadata)
+
+    def run(self):
+        """
+        The main function of the task class, which will be called by the future asynchronously
+        """
+        remaining_t = self.estimate_total_duration()
+        acquired_data = []
+        # iterate through streams
+        for stream in self._streams:
+            zstack = []
+            # for each stream, iterate through zlevels (if exist)
+            for i, z in enumerate(self._zlevels[stream]):
```

Comment:
The specification was much more fancy in terms of zlevels meaning. At least, I'd say that it should support the stream not being in zlevels. In this case, the stream is acquired just once.

Reply:
I did not get the point. Could you please elaborate?

Reply:
The spec says "zlevels (dict Actuator → list of floats, or dict Stream → list of floats): In both cases, the list of float is the positions of the z, in order. They must be spread regularly.  If Actuator → list, there should be only one entry. The actuator is used to move the focus, it should have a “z” axis. In such case, for each zlevel in the list, all the streams are acquired. If it’s Stream → list, each stream is acquired serially. For the streams which are in the zlevels dict, the .focuser of the stream is used to change to each zlevel designated. **The streams which are not in the zlevels dict are acquired only once.** (In other words, the FM streams are acquired at each zlevel, and the SEM streams are acquired only once). "

So you have to be extra careful at this line, as `zelevels` may not contain the stream.

Reply:
Indeed, i made some changes. Now the SEM are not included in the zlevels dictionary. Please have a look at the latest commits